pkgs/native_assets_builder/doc/protocol/v1.md

```diff
@@ -0,0 +1,216 @@
+#
+# Remark: Avoid using the name "build_output.yaml" for an asset file, this is
+#         forbidden.
+out_dir: /absolute/path/to/out_dir/
+
+# Name of the package that contains the `build.dart`
+#
+# Remark: This is entirely redundant since this is a config file specified to
+# `build.dart`, and the author of `build.dart` probably knows the name of the
+# package they are writing.
+package_name: my_package_with_native_assets
+
+# Path to root folder for the package that contains `build.dart`.
+#
+# This is useful if `build.dart` wishes to find source code files embedded in
+# its own package and compile them to an asset.
+#
+# Note that this will be most likely a path in the pub cache when the package
+# with a `build.dart` is a dependency of another package.
+package_root: /absolute/path/to/my_package_with_native_assets
+
+# Target architecture
+#
+# Combined with `target_os` this specifies the "target" for which assets
+# should be built.
+#
+# Not provided on dry runs.
+target_architecture: x64 | ia32 | arm | arm64 | riscv32 | riscv64
+
+# Target operating system
+#
+# Combined with `target_architecture` this specifies the "target" for which
+# assets should be built.
+target_os: android | ios | linux | macos | windows
+
+# Version of this file.
+version: 1.0.0
+```
+
+### `build_output.yaml`
+Mapping from `assetId`s to files created by `build.dart`.
+
+```yaml
+# The list of assets.
+#
+# In dry runs, must contain assets for each architecture for the requested os.
+assets:
+  - id: 'package:my_package_with_native_assets/src/foo.dart'
+    link_mode: dynamic | static | prefer-dynamic | prefer-static
+    path:
+      path_type: absolute | system | process | executable
+      # Only provided for path_type absolute and system.
+      #
+      # If path_type absolute: The absolute path to the file name.
+      #
+      # If path_type system: The path of the dynamic library as available on 
+      # the target machine's PATH.
+      uri: /absolute/path/to/outdir/arbitrary_filename.whatever
+    target: linux_x64
+  ...
+
+# The files used by this build.
+#
+# If any of the files in [dependencies2] are modified after [timestamp], the
+# build will be re-run.
+#
+# Not output on dry runs.
+dependencies:
+  - /absolute/path/to/my_package_with_native_assets/build.dart
+  ...
+
+# The time the build this output is for started.
+#
+# Must be before any of the files in dependencies are read.
+timestamp: 2024-01-02 17:05:35.000
+
+# Metadata usable for build.dart of packages directly depending on this package.
+#
+# Not output in dry runs.
+metadata:
+  # Key value pairs.
+  some_key: some_value
+
+# Version of this file.
+version: 1.0.0
+```
+
+
+### `.dart_tool/native_assets.yaml`
+
+This file is internal to the Dart SDK and Flutter SDK.
+It produced by `native_asset_builder` (as embedded in the SDKs).
+For each `target` it maps from `assetId` to `url`
+(with `urlKind` indicating if it's a relative or absolute path).
```

Comment:
Again separate implementation from contents - what is `urlKind`?

Reply:
I'm not sure that we should have documentation about `.dart_tool/native_assets.yaml` in here. It's an internal vm format that happens to have some similarities with the native code assets.

See below, it's the path_type. And maybe that should have a different name. I'm open to suggestions. In the context of this file it is:

* absolute paths are absolute paths on the system where the Dart executable is running. This path type only really makes sense within the context of running things in JIT mode on a developers' machine, or in an Android app where the root path is the Android bundle.
* relative paths are relative to the kernel or aot snapshot. This makes sense if you're shipping a bundle through `dart build` and your asset is placed in a folder relative to the aotsnapshot.
* system paths are the same as for the protocol above.
* process "paths" have no path, and are the same as for the protocol above. 
* executable "paths" have no path, and are the same as for the protocol above.